src/providers/tfc/planProvider.ts

```diff
@@ -0,0 +1,424 @@
+
+    const items: vscode.TreeItem[] = [];
+    if (planLog && planLog.plannedChanges) {
+      items.push(new PlannedChangesItem(planLog.plannedChanges, planLog.changeSummary));
+    }
+    if (planLog && planLog.driftChanges) {
+      items.push(new DriftChangesItem(planLog.driftChanges, planLog.driftSummary));
+    }
+    //  && planLog.outputs.size > 0
+    if (planLog && planLog.outputs) {
+      items.push(new OutputsItem(planLog.outputs));
+    }
+    if (planLog && planLog.diagnostics && planLog.diagnosticSummary && planLog.diagnostics.length > 0) {
+      items.push(new DiagnosticsItem(planLog.diagnostics, planLog.diagnosticSummary));
+    }
+    return items;
+  }
+
+  private async getPlanFromUrl(plan: PlanTreeItem): Promise<PlanLog | undefined> {
+    const session = await vscode.authentication.getSession(TerraformCloudAuthenticationProvider.providerID, [], {
+      createIfNone: false,
+    });
+
+    if (session === undefined) {
+      return;
+    }
+
+    try {
+      const result = await axios.get(plan.logReadUrl, {
+        headers: { Accept: 'text/plain' },
+        responseType: 'stream',
+      });
+      const lineStream = readline.createInterface({
+        input: result.data,
+        output: new Writable(),
+      });
+
+      const planLog: PlanLog = {};
+
+      for await (const line of lineStream) {
+        try {
+          const logLine: LogLine = JSON.parse(line);
+
+          if (logLine.type === 'planned_change' && logLine.change) {
+            if (!planLog.plannedChanges) {
+              planLog.plannedChanges = [];
+            }
+            planLog.plannedChanges.push(logLine.change);
+            continue;
+          }
+          if (logLine.type === 'resource_drift' && logLine.change) {
+            if (!planLog.driftChanges) {
+              planLog.driftChanges = [];
+            }
+            if (!planLog.driftSummary) {
+              planLog.driftSummary = {
+                changed: 0,
+                deleted: 0,
+              };
+            }
+            planLog.driftChanges.push(logLine.change);
+            if (logLine.change.action === 'update') {
+              planLog.driftSummary.changed += 1;
+            }
+            if (logLine.change.action === 'delete') {
+              planLog.driftSummary.deleted += 1;
+            }
+            continue;
+          }
+          if (logLine.type === 'change_summary' && logLine.changes) {
+            planLog.changeSummary = logLine.changes;
+            continue;
+          }
+          if (logLine.type === 'outputs' && logLine.outputs) {
+            // JSON.parse() won't do the conversion from [JSON] object
+            // to the Map so we have to do it here.
+            Object.entries(logLine.outputs).forEach(function (value: [string, OutputChange]) {
+              if (!planLog.outputs) {
+                planLog.outputs = new Map();
+              }
+              planLog.outputs.set(value[0], value[1]);
+            });
```

Comment:
FWIW This is my understanding of the situation - if that's wrong or if there is any more elegant way of dealing with this I'm all ears!

Reply:
Usually `JSON.parse` will create an object here and you can use it very similar to a map. Why do you prefer a map?

Reply:
There may be a little bit of a hesitancy due to me coming to this from Go, where it's difficult to loop over objects, for mostly good reasons like meaning of the data type and the way its stored in memory etc. I do keep forgetting though that things which are wrong in one language could be business as usual in another. 😅 

Even if looping over object is fine in JS I still think there is still benefit to the IntelliSense gained through this explicit map-based type and possibly validation of the object structure as it comes from the API, is there not?

Reply:
Don't get me wrong, I think using maps can bring great benefits over using objets in JS.

But I'm not sure I'm following here. `JSON.parse` returns an object and you loop over an object (`Object.entries(logLine.outputs).forEach`). If `logLine.outputs` were a map, as the types (and IntelliSense) suggest, you could use `logLine.outputs.forEach` to iterate over the contents. But this will throw an error at runtime.

By assigning a type to logLine (`const logLine: LogLine`), _we are telling_ TypeScript that this result is of type `LogLine`, but we don't validate that claim. (This is why we use zod for the other TFC API responses, to make sure they really do conform to a schema). But using zod for the log response may be overkill.
The correct type for `outputs` in `LogLine` would be an object:
```js
export interface LogLine {
  ...
  outputs?: {
    [key: string]: {
      action: ChangeAction;
      sensitive: boolean;
    };
  };
}
```